src/download.rs

```diff
@@ -0,0 +1,270 @@
+
+            // fail
+            return Err(err);
+        }
+
+        // report the output file path
+        println!("{}", path.display());
+    }
+
+    Ok(())
+}
+
+// Check an existing image and signature for validity.  The image cannot
+// have been decompressed after downloading.  Return an error if invalid for
+// any reason.
+fn check_image_and_sig(source: &ImageSource, path: &Path, sig_path: &Path) -> Result<()> {
+    // ensure we have something to check
+    if source.signature.is_none() {
+        return Err("no signature available; can't check existing file".into());
+    }
+    let signature = source.signature.as_ref().unwrap();
+
+    // compare signature to expectation
+    let mut sig_file = OpenOptions::new()
+        .read(true)
+        .open(sig_path)
+        .chain_err(|| format!("opening {}", sig_path.display()))?;
+    let mut buf = Vec::new();
+    sig_file
+        .read_to_end(&mut buf)
+        .chain_err(|| format!("reading {}", sig_path.display()))?;
+    if &buf != signature {
+        return Err("signature file doesn't match source".into());
+    }
+
+    // open image file
+    let mut file = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .chain_err(|| format!("opening {}", path.display()))?;
+
+    // perform GPG verification
+    let mut verify_reader = GpgReader::new(&mut file, signature)?;
+    let mut buf: [u8; 4096] = [0; 4096];
+    while verify_reader
+        .read(&mut buf)
+        .chain_err(|| "reading image file")?
+        > 0
+    {}
+
+    Ok(())
+}
+
+/// Copy the image to disk, and also the signature if appropriate.
+fn write_image_and_sig(
+    source: &mut ImageSource,
+    path: &Path,
+    sig_path: &Path,
+    decompress: bool,
+) -> Result<()> {
+    // open output
+    let mut dest = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(path)
+        .chain_err(|| format!("opening {}", path.display()))?;
+
+    // download and verify image
+    write_image(source, &mut dest, decompress)?;
+
+    // write signature, if relevant
+    if !decompress && source.signature.is_some() {
+        let mut sig_dest = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(sig_path)
+            .chain_err(|| format!("opening {}", sig_path.display()))?;
+        sig_dest
+            .write_all(source.signature.as_ref().unwrap())
+            .chain_err(|| "writing signature data")?;
+    }
+
+    Ok(())
+}
+
+/// Copy the image to disk and verify its signature.
+pub fn write_image(source: &mut ImageSource, dest: &mut File, decompress: bool) -> Result<()> {
+    // wrap source for GPG verification
+    let mut verify_reader: Box<dyn Read> = {
+        if let Some(signature) = source.signature.as_ref() {
+            Box::new(GpgReader::new(&mut source.reader, signature)?)
+        } else {
+            Box::new(&mut source.reader)
```

Comment:
I have some (very minor) distress about this `&mut` and the one above, as they cause a long chain of upwards `mut` locations, only because here we want to read through the reader.
Do you think it could be feasible to avoid that by having an additional wrapping like `Cursor` or `BufReader`?

Reply:
I'm not sure how?  `Read.read()` fundamentally expects `&mut self`.  We could get around that with a `RefCell` but that doesn't seem like the right approach.  Or am I misunderstanding your point?

Reply:
I'm not totally sure (so it's fine to postpone and I'll experiment a bit, later on).
My gut feeling was that `Cursor` (or something akin) could provide the mutability for `Read` while keeping the inner type immutable.

Reply:
I looked into it a bit.  `Cursor` only wraps a `[u8]`, and `BufReader` expects an inner `Read`.

Reply:
Ack. Then I think I'm thinking about something not directly possible. Let's keep this as it is.